R/iterate-helpers.R

```diff
@@ -0,0 +1,115 @@
+#'   `?page=2`, or `?offset=1`, `offset=21`.
+#' * `iterate_with_cursor()` updates a query parameter with the value of a
+#'    cursor found somewhere in the response.
+#'  * `iterate_with_link_url()` follows the url found the `Link` header.
+#'    See `resp_link_url()` for more details.
+#'
+#' @param param_name Name of query parameter.
+#' @param start Starting value.
+#' @param offset Offset for each page.
+#' @param resp_complete A callback function that takes a response (`resp`)
+#'   and returns `TRUE` if there are no further pages.
+#' @param resp_pages A callback function that takes a response (`resp`) and
+#'   returns the total number of pages, or `NULL` if unknown. It will only
+#'   be called once.
+#' @export
+#' @examples
+#' req <- request(example_url()) |>
+#'   req_url_path("/iris") |>
+#'   req_throttle(10) |>
+#'   req_url_query(limit = 50)
+#'
+#' # If you don't know the total number of pages in advance, you can
+#' # provde a `resp_complete()` callback
+#' is_complete <- function(resp) {
+#'   length(resp_body_json(resp)$data) == 0
+#' }
+#' resps <- req_perform_iteratively(
+#'   req,
+#'   next_req = iterate_with_offset("page_index", resp_complete = is_complete),
+#'   max_reqs = Inf
+#' )
+#'
+#' \dontrun{
+#' # Alternatively, if the response returns the total number of pages (or you
+#' # can easily calculate it), you can use the `resp_pages()` callback which
+#' # will generate a better progress bar.
+#'
+#' resps <- req_perform_iteratively(
+#'   req %>% req_url_query(limit = 1),
+#'   next_req = iterate_with_offset(
+#'     "page_index",
+#'     resp_pages = function(resp) resp_body_json(resp)$pages
+#'   ),
+#'   max_reqs = Inf)
+#' }
+#'
+iterate_with_offset <- function(param_name,
+                                start = 1,
+                                offset = 1,
+                                resp_pages = NULL,
+                                resp_complete = NULL) {
+  check_string(param_name)
+  check_number_whole(start)
+  check_number_whole(offset, min = 1)
+  check_function2(resp_pages, args = "resp", allow_null = TRUE)
+  check_function2(resp_complete, args = "resp", allow_null = TRUE)
+  resp_complete <- resp_complete %||% function(resp) FALSE
+
+  known_total <- FALSE
+  i <- start # assume already fetched
+
+  function(resp, req) {
+    if (!is.null(resp_pages) && !known_total) {
+      n <- resp_pages(resp)
+      if (!is.null(n)) {
+        known_total <<- TRUE
+        signal("", class = "httr2_total_pages", n = n)
+      }
+    }
+
+    if (!isTRUE(resp_complete(resp))) {
+      i <<- i + offset
+      req %>% req_url_query(!!param_name := i)
+    }
+  }
+}
+
+#' @rdname iterate_with_offset
+#' @export
+#' @param resp_param_value A callback function that takes a response (`resp`)
+#'   and returns the next cursor value. Return `NULL` if there are no further
+#'   pages.
+iterate_with_cursor <- function(param_name, resp_param_value) {
```

Comment:
Is there a reason to not include a `resp_pages` argument?

Reply:
I didn't include a `resp_pages()` for either of these two helpers because I think it's relatively rare that the total number of pages is available with these interfaces. But I guess there's no harm in adding it.